hazardous/metrics/_concordance_index.py

```diff
@@ -0,0 +1,471 @@
+import warnings
+from collections import Counter, defaultdict
+
+import numpy as np
+from scipy.interpolate import interp1d
+
+from .._ipcw import KaplanMeierIPCW
+from ..utils import check_y_survival
+
+
+def concordance_index_incidence(
+    y_test,
+    y_pred,
+    y_train=None,
+    ipcw_estimator="km",
+    time_grid=None,
+    taus=None,
+    event_of_interest=1,
+    tied_tol=1e-8,
+):
+    r"""Time-dependent concordance index for prognostic models with competing risks \
+    using inverse probability of censoring weighting.
+
+    .. math::
+
+        \mathrm{C}(t) = \frac{\sum_{i=1}^n \sum_{j=1}^n (\tilde{A}_{ij}
+        \hat{W}_{ij, 1}^{-1} + \tilde{B}_{ij} \hat{W}_{ij, 2}^{-1}) Q^{ij}(t)
+        \tilde{N}^1_i(t)}
+        {\sum_{i=1}^n \sum_{j=1}^n (\tilde{A}_{ij}
+        \hat{W}_{ij, 1}^{-1} + \tilde{B}_{ij} \hat{W}_{ij, 2}^{-1}) \tilde{N}^1_i(t)}
+
+    where:
+
+    .. math::
+
+        \begin{align}
+        \tilde{N}^1_i(t) &= I\{\tilde{T}_i \leq t, \tilde{D}_i = 1\} \\
+        \tilde{A}_{ij} &= I\{\tilde{T}_i < \tilde{T}_j \cup (\tilde{T}_i =
+        \tilde{T}_j \cap D_j = 0)\} \\
+        \tilde{B}_{ij} &= I\{\tilde{T}_i \geq \tilde{T}_j, D_j = 2\} \\
+        \hat{W}_{ij,1} &= \hat{G}(\tilde{T}_i-|X_i) \hat{G}(\tilde{T}_i|X_j) \\
+        \hat{W}_{ij,2} &= \hat{G}(\tilde{T}_i-|X_i) \hat{G}(\tilde{T}_j-|X_j) \\
+        Q_{ij}(t) &= I\{M(t, X_i) > M(t, X_j)\}
+        \end{align}
+
```

Comment:
it's probably defined elsewhere, but on standalone reading, might be useful to redefine $\tilde{T}$ and $\tilde{D}$.